### PR TITLE
PLATFORM-7909 - Upgrade Cheevos to 1.39

### PR DESCRIPTION
**Pull request by Michal-Chatlas1221**
Changelog:
* Linting with `phpcsf`
* Removing execute property from all files (`minus-x fix .`)
* Repackaging classes to use single `AutoloadNamespaces` in `extension.json`
* Update Special pages to inject services via constructor
* Use HookHandler for hooks
* Removing obsolete / unused maintenance scripts
* Migrating from `Cheevos` class to `AchievementService`, `FriendService` and `CheevosClient` classes

Left for future:
* Refactor of `CheevosModel` classes - stop using `__call`, `__get` and `__set` magic methods.
Use class properties instead array of values
* Further refactor of `Template` classes

### CODE REVIEW

File: src/CheevosHooks.php [hunk 1]
```diff
@@ -710,367 +708,249 @@ public static function broadcastAchievement( CheevosAchievement $achievement, st
...
-		$db = MediaWikiServices::getInstance()->getDBLoadBalancer()->getConnection( DB_PRIMARY );
-		$actorQuery = [ 'tables' => [], 'joins' => [] ];
-		$userField = 'rev_user';
-
-		$db->select( [ 'revision' ] + $actorQuery['tables'], [ 'count(*)' ], [], __METHOD__, [
-			'GROUP BY' => $userField,
-			'SQL_CALC_FOUND_ROWS',
-		], $actorQuery['joins'] );
-		$calcRowsResult = $db->query( 'SELECT FOUND_ROWS() AS rowcount;' );
-		$total = $calcRowsResult->fetchRow();
```
Comment (mmeller-wikia) [inline src/CheevosHooks.php]: Double-check on CR
It is replaced by:
```
$contributorCount = $this->loadBalancer->getConnection( DB_REPLICA )
			->selectRowCount(
				'revision',
				'distinct(rev_actor)',
				[],
				__METHOD__
			);

File: src/CheevosHelper.php [hunk 2]
```diff
@@ -12,14 +12,83 @@
...
+
+	public function __construct( private AchievementService $achievementService ) {
+	}
+
+	public function increment( string $stat, int $delta, UserIdentity $user, array $edits = [] ): void {
+		// Register shutdown function to actually save increments
+		if ( !self::$shutdownRegistered && PHP_SAPI !== 'cli' ) {
+			self::$shutdownRegistered = true;
+			register_shutdown_function( fn() => $this->doIncrements() );
+		}
```
Comment (mmeller-wikia) [inline src/CheevosHelper.php:38]: Let's register shutdown function only when required - when `increment` function has been called

File: extension.json [hunk 3]
```diff
@@ -225,40 +225,41 @@
 				"LinkRenderer",
 				"DBLoadBalancer",
 				"RedisCache",
+				"Cheevos\\AchievementService",
 				"Cheevos\\CheevosHelper"
 			]
 		}
 	},
 	"Hooks": {
-		"ArticleDeleteComplete": "Cheevos\\CheevosHooks::onArticleDeleteComplete",
```
Comment (mmeller-wikia) [inline extension.json]: Replaced by `PageDeleteComplete`

File: src/CheevosHooks.php [hunk 4]
```diff
@@ -237,304 +251,108 @@ public static function onPageSaveComplete(
...
-		}
-		catch ( CheevosException $e ) {
+
+			$editCount = (int)( $stats[$fromUser->getId()]['article_edit']['count'] ?? 0 );
+			return $editCount >= $editsToComment;
+		} catch ( CheevosException $e ) {
 			wfDebug( "Encountered Cheevos API error getting article_edit count." );
+			// TODO--is it a good idea to allow on error??
+			return true;
 		}
```
Comment (mmeller-wikia) [inline src/CheevosHooks.php:281]: Is it a good idea: allowing posting comments on CurseProfiles on errors

File: maintenance/CompSubscriptions.php [hunk 5]
```diff
@@ -1,87 +0,0 @@
...
-
-namespace Cheevos;
-
-require_once dirname( __DIR__, 3 ) . '/maintenance/Maintenance.php';
-
-use Cheevos\Points\PointsCompReport;
-use ExtensionRegistry;
-use Maintenance;
-
-class CompSubscriptions extends Maintenance {
```
Comment (mmeller-wikia) [inline maintenance/CompSubscriptions.php:19]: Same functionality is available via `SpecialPointsComp`

File: maintenance/DumpWikiPointsSqlImport.php [hunk 6]
```diff
@@ -1,150 +0,0 @@
...
- * @license   GPL-2.0-or-later
- * @link      https://www.gamepedia.com/
- */
-
-require_once __DIR__ . '/../../../maintenance/Maintenance.php';
-
-use Cheevos\CheevosHelper;
-use MediaWiki\MediaWikiServices;
-
-class DumpWikiPointsSqlImport extends Maintenance {
```
Comment (mmeller-wikia) [inline maintenance/DumpWikiPointsSqlImport.php:18]: WikiPoints tables will be dropped with upgrade

File: extension.json [hunk 7]
```diff
@@ -22,28 +21,6 @@
...
-			"award_achievements": true,
-			"delete_achievements": true,
-			"edit_achievements": true,
-			"edit_meta_achievements": true,
-			"restore_achievements": true,
-			"wiki_points_admin": true,
-			"wpa_adjust_points": true,
-			"points_comp_reports": true
-		},
-		"hydra_admin": {
```
Comment (mmeller-wikia) [inline extension.json:37]: Removed in `LocalPermissionOverrides::removeUnusedGroups` in `SetupAfterCache` hook

**Comment by slayful** [comment]
Let's incorporate these changes https://github.com/Wikia/mediawiki-extensions-HydraCore/pull/11

**Review by maaartyyynaa — COMMENTED** [review]
I think we should split PHPUnit / PHPCS / Phan related changes into separate PR, from the changes needed for the MW upgrade. At this point this PR is hard to review.

Is there a way to do that?

**Comment by mmeller-wikia** [comment]
Split into https://github.com/Wikia/mediawiki-extensions-Cheevos/pull/11
https://github.com/Wikia/mediawiki-extensions-Cheevos/pull/12
https://github.com/Wikia/mediawiki-extensions-Cheevos/pull/13
https://github.com/Wikia/mediawiki-extensions-Cheevos/pull/14
I may try to rebase some commits, but it will be much harder than splitting existing commits

File: src/CheevosHooks.php [hunk 8]
```diff
@@ -413,27 +414,19 @@ public function onParserGetVariableValueSwitch( $parser, &$variableCache, $magic
...
-		$redis = $this->redisCache->getConnection( 'cache' );
-
-		$redisKey = 'cheevos:contributors:' . CheevosHelper::getSiteKey();
-		if ( $redis ) {
-			$cache = $redis->get( $redisKey );
-			if ( $cache !== false ) {
-				return (string)$cache;
+		$method = __METHOD__;
+		return $this->objectCache->getWithSetCallback(
+			$this->objectCache->makeKey( 'cheevos', 'contributors' ),
```
Comment (mmeller-wikia) [inline src/CheevosHooks.php:419]: Use per wiki cache instead global one

File: src/Points/PointsDisplay.php [hunk 9]
```diff
@@ -186,16 +185,11 @@ public static function pointsBlockHtml(
...
-			$redis = MediaWikiServices::getInstance()
-				->getService( RedisCache::class )
-				->getConnection( 'cache' );
-			if ( $redis !== false ) {
-				foreach ( $siteKeys as $siteKey ) {
-					if ( !empty( $siteKey ) ) {
-						$wiki = CheevosHelper::getWikiInformation( $siteKey );
-						if ( !empty( $wiki ) ) {
-							$wikis[$siteKey] = $wiki;
-						}
```
Comment (mmeller-wikia) [inline src/Points/PointsDisplay.php]: `$redis` is effectively unused in this code

File: src/AchievementService.php [hunk 10]
```diff
@@ -56,7 +56,7 @@ public function broadcastAchievement( CheevosAchievement $achievement, string $s
 	/** Invalidate API Cache */
 	public function invalidateCache(): void {
 		$redis = $this->redisCache->getConnection( self::REDIS_CONNECTION_GROUP );
-		if ( $redis === false ) {
+		if ( !$redis ) {
 			return;
 		}
```
[the remark this answers is not in the document]
Comment (Michal-Chatlas1221) [reply src/AchievementService.php]: maybe we can use `!empty( $redis ) `

**Review by Michal-Chatlas1221 — COMMENTED** [review]
I can't approve since I'm still technically considered author by github

**Review by mmeller-wikia — APPROVED** [review]

**Comment by mmeller-wikia** [comment]
> I can't approve since I'm still technically considered author by github

I approved this PR as per [Michal-Chatlas1221](https://github.com/Michal-Chatlas1221) comment

**Pull request merged**